File: api/BeSureApi/Models/ManPower.cs
namespace BeSureApi.Models
{
    public class ManPowerCreate
    {
        public int ContractId { get; set; }
        public int CustomerSiteId { get; set; }
        public int MspLocationId { get; set; }
        public int DistanceToCustomerSite { get; set; }
        public int DistanceUnitId { get; set; }
        public int EngineerTypeId { get; set; }
        public int EngineerLevelId { get; set; }
        public int BudgetedCost { get; set; }
        public int EmployeeId { get; set; }
        public DateTime EmployeeReportedOn { get; set; }
        public int EmployeeBillingRate { get; set; }
    }

    public class ManPowerSummaryCreate
    {
        public int ContractId { get; set; }
        public int CustomerSiteId { get; set; }
        public int TenantOfficeInfoId { get; set; }
        public int EngineerTypeId { get; set; }
        public int EngineerLevelId { get; set; }
        public decimal EngineerMonthlyCost { get; set; }
        public decimal EngineerCount { get; set; }
        public decimal DurationInMonth { get; set; }
        public decimal CustomerAgreedAmount { get; set; }
        public string? Remarks { get; set; }
        public int CreatedBy { get; set; }
    }

    public class ManPowerSummaryList
    {
        public int Id { get; set; }
        public string CustomerSite { get; set; }
        public string TenantOffice { get; set; }
        public string EngineerType { get; set; }
        public string EngineerLevel { get; set; }
        public int EngineerMonthlyCost { get; set; }
        public int EngineerCount { get; set; }
        public int DurationInMonth { get; set; }
        public decimal BudgetedAmount { get; set; }
        public decimal CustomerAgreedAmount { get; set; }
        public decimal MarginAmount { get; set; }
        public string? Remarks { get; set; }
    }

    public class ManPowerList
    {
        public int Id { get; set; }
        public int ContractId { get; set; }
        public string CustomerSiteName { get; set; }
        public string MspLocationName { get; set; }
        public int CustomerSiteId { get; set; }
        public int MspLocationId { get; set; }
        public int EmployeeId { get; set; }
        public int DistanceToCustomerSite { get; set; }
        public int DistanceUnitId { get; set; }
        public int EngineerTypeId { get; set; }
        public int EngineerLevelId { get; set; }
        public int BudgetedCost { get; set; }
        public int CreatedBy { get; set; }
        public DateTime CreatedOn { get; set; }
        public string EmployeeName { get; set; }
        public DateTime EmployeeReportedOn { get; set; }
        public int EmployeeBillingRate { get; set; }
        public DateTime EmployeeLastSiteWorkingDate { get; set; }
        public string CreatedUserName { get; set; }
        public int ModifiedBy { get; set; }
        public DateTime ModifiedOn { get; set; }
        public string ModifiedUserName { get; set; }
        public Boolean IsDeleted { get; set; }
    }

    public class ManPowerUpdate
    {
        public int Id { get; set; }
        public int ContractId { get; set; }
        public int CustomerSiteId { get; set; }
        public int MspLocationId { get; set; }
        public int DistanceToCustomerSite { get; set; }
        public int DistanceUnitId { get; set; }
        public int EngineerTypeId { get; set; }
        public int EngineerLevelId { get; set; }
        public int BudgetedCost { get; set; }
        public int EmployeeId { get; set; }
        public int EmployeeBillingRate { get; set; }
    }

    public class ManPowerSummaryDetails
    {
        public int Id { get; set; }
        public int ContractId { get; set; }
        public int CustomerSiteId { get; set; }
        public int TenantOfficeInfoId { get; set; }
        public int EngineerTypeId { get; set; }
        public int EngineerLevelId { get; set; }
        public decimal EngineerMonthlyCost { get; set; }
        public decimal EngineerCount { get; set; }
        public decimal DurationInMonth { get; set; }
        public decimal CustomerAgreedAmount { get; set; }
        public string? Remarks { get; set; }
    }

    public class ManPowerSummaryUpdate
    {
        public int Id { get; set; }
        public int ContractId { get; set; }
        public int CustomerSiteId { get; set; }
        public int TenantOfficeInfoId { get; set; }
        public int EngineerTypeId { get; set; }
        public int EngineerLevelId { get; set; }
        public decimal EngineerMonthlyCost { get; set; }
        public decimal EngineerCount { get; set; }
        public decimal DurationInMonth { get; set; }
        public decimal CustomerAgreedAmount { get; set; }
        public string? Remarks { get; set; }
        public int ModifiedBy { get; set; }
    }
    public class IsAgreedAmountExceeded
    {
        public bool IsValueExceeded { get; set; }
    }
    public class IsSummaryAmountExceeded
    {
        public bool IsAmcValueExceeded { get; set; }
        public bool IsFmsValueExceeded { get; set; }

    }
}

File: ui/src/types/manpower.ts
import { array, number, Decoder, nullable, object, string,iso8601, boolean } from 'decoders';

export interface Configurations{
  EngineerType: Select[],
  EngineerLevel: Select[],
  }

export interface Select {
  value: any,
  label: any
}

export interface SelectMasterDataDetails {
  Select: Select[]
}


  export interface EntitiesDetails {
    Id: number;
    Name: string;
  } 

  export interface CustomerSites{
    Id:number;
    SiteName: string;
  }
  
  export interface EngineersDetails {
    Id: number;
    FullName: string;
  }

export interface ManpowerSummaryCreation {
  ContractId: string;
  CustomerSiteId: number | string;
  TenantOfficeInfoId: number | string;
  EngineerMonthlyCost: number | string;
  EngineerCount: number | string;
  DurationInMonth: number | string;
  EngineerTypeId: number | string;
  EngineerLevelId: number | string;
  CustomerAgreedAmount: number | string;
  Remarks: string | null;
}

export interface EntitiesDetails {
  Id: number;
  Name: string;
}

export interface ManPowerEditTemplate {
  Id: number | string;
  ContractId: string | number;
  CustomerSiteId: number | string;
  TenantOfficeInfoId: number | string;
  EngineerMonthlyCost: number | string;
  EngineerCount: number | string;
  DurationInMonth: number | string;
  EngineerTypeId: number | string;
  EngineerLevelId: number | string;
  CustomerAgreedAmount: number | string;
  Remarks: string | null;
}

export const  manpowerEditTemplateDecoder: Decoder<ManPowerEditTemplate> = object({
  Id:number,
  ContractId:number,
  TenantOfficeInfoId: number,
  EngineerMonthlyCost: number,
  CustomerSiteId : number,
  EngineerCount : number,
  DurationInMonth : number,
  EngineerTypeId: number,
  CustomerAgreedAmount: number,
	EngineerLevelId:number,
  Remarks: nullable(string),
});


export interface SelectedManPowerDetails {
  ContractManpowerSummary: ManPowerEditTemplate;
}

export const selectedManPowerDetailsDecoder: Decoder<SelectedManPowerDetails> = object({
  ContractManpowerSummary: manpowerEditTemplateDecoder,
});

export interface ManPowerStatusTemplate {
  Id:number;
  CustomerSiteName : string;
  MspLocationName : string;
  DistanceToCustomerSite: string;
  DistanceUnitId :  number;
  EngineerTypeId :  number ;
  EngineerLevelId :  number;
  BudgetedCost  :  string;
  EmployeeName :  string;
  EmployeeReportedOn : typeof iso8601 | string;
  EmployeeBillingRate : string;
  IsDeleted:Boolean;
}

export interface ManPowerEditResponse {
  IsUpdated: Boolean; 
}

export const manPowerEditDecoder: Decoder<ManPowerEditResponse> = object({
  IsUpdated: boolean,
});

export interface ManpowerSummaryList {
  Id: number;
  CustomerSite: string;
  CustomerAgreedAmount: number;
  EngineerType: string;
  EngineerLevel: string;
  EngineerCount: number;
  DurationInMonth: number;
  TenantOffice: string;
  EngineerMonthlyCost: number;
  Remarks: string | null;
  MarginAmount: number;
  BudgetedAmount: number;
}

export const manpowerSummaryListDecoder: Decoder<ManpowerSummaryList> = object({
  Id: number,
  CustomerSite: string,
  CustomerAgreedAmount:number,
  EngineerType: string,
  EngineerLevel: string,
  EngineerCount: number,
  DurationInMonth: number,
  TenantOffice: string,
  EngineerMonthlyCost: number,
  Remarks: nullable(string),
  MarginAmount: number,
  BudgetedAmount: number
});

export interface ContractManpowerSummaryList {
  ContractManPowerSummaryList: ManpowerSummaryList[];
  TotalRows: number;
  PerPage:number;
}

export const contractManpowerSummaryListDecoder: Decoder<ContractManpowerSummaryList> = object({
  ContractManPowerSummaryList: array(manpowerSummaryListDecoder),
  TotalRows: number,
  PerPage:number
});